### PR TITLE
graph: Validate @derivedFrom directives on subgraph schemas

### PR DESCRIPTION
**Pull request by lutter**
These validations follow the corresponding validations in graph-cli, and make sure users do not deploy invalid subgraphs, e.g., because they have not updated their subgraph or graph-cli

Fixes https://github.com/graphprotocol/graph-node/issues/1213

### CODE REVIEW

File: graph/src/data/graphql/validation.rs [hunk 1]
```diff
@@ -93,10 +100,199 @@ pub fn get_object_type_definitions(schema: &Document) -> Vec<&ObjectType> {
...
+            reason.to_owned(),
+        )
+    }
+
+    let type_definitions = get_object_type_definitions(schema);
+    let object_and_interface_type_fields = get_object_and_interface_type_fields(schema);
+
+    // Iterate over all derived fields in all entity types; include the
+    // `field` argument of @derivedFrom directive
+    for (object_type, field, target_field) in type_definitions
```
Comment (Jannis) [inline graph/src/data/graphql/validation.rs]: We should probably move the schema validation code into the `graph-graphql` crate. Like you already mentioned, a number of helpers have already been added there but can't be used. This is another: `graph_graphql::schema::ast::get_derived_from_directive` and `graph_graphql::schema::ast::get_derived_from_field` already exist there.

Comment (Jannis) [reply graph/src/data/graphql/validation.rs]: We don't have to do it now though.

File: graph/src/data/graphql/validation.rs [hunk 2]
```diff
@@ -93,10 +100,199 @@ pub fn get_object_type_definitions(schema: &Document) -> Vec<&ObjectType> {
...
+        .clone()
+        .iter()
+        .map(|object_type| {
+            object_type
+                .fields
+                .iter()
+                .map(move |field| (object_type, field))
+        })
+        .flatten()
```
Comment (Jannis) [inline graph/src/data/graphql/validation.rs]: The above `map` and `flatten` could be combined with `flat_map`.

File: graph/src/data/graphql/validation.rs [hunk 3]
```diff
@@ -93,10 +100,199 @@ pub fn get_object_type_definitions(schema: &Document) -> Vec<&ObjectType> {
...
+        .filter_map(|(object_type, field)| {
```
Comment (Jannis) [inline graph/src/data/graphql/validation.rs]: The `if let Some` logic here could be simplified a little by using `map`:
```rust
...
    .filter_map(|(object_type, field)| {
        field
            .directives
            .iter()
            .find(|directive| directive.name == "derivedFrom")
            .map(|directive| {
                (
                    object_type,
                    field,
                    directive
                        .arguments
                        .iter()
                        .find(|(name, _)| name == "field")
                        .map(|(_, value)| (object_type, field, target_field))
                )
            })
    })
```

File: graph/src/data/graphql/validation.rs (hunk 1 above)
Comment (Jannis) [reply graph/src/data/graphql/validation.rs]: The alternative would be to make the `graph_graphql` crate unaware of an Graph-specific data structures and make it a dependency of `graph`.

File: graph/src/data/graphql/validation.rs [hunk 4]
```diff
@@ -93,10 +100,199 @@ pub fn get_object_type_definitions(schema: &Document) -> Vec<&ObjectType> {
...
+
+        // Check that the type we are deriving from exists
+        let target_type_name = get_base_type(&field.field_type);
+        let target_fields = object_and_interface_type_fields
+            .get(target_type_name)
+            .ok_or_else(|| {
+                invalid(
+                    object_type,
+                    &field.name,
+                    "the type of the field must be an existing entity type",
```
Comment (Jannis) [inline graph/src/data/graphql/validation.rs]: I'd change this to `entity or interface type`.

File: graph/src/data/graphql/validation.rs [hunk 5]
```diff
@@ -93,10 +100,199 @@ pub fn get_object_type_definitions(schema: &Document) -> Vec<&ObjectType> {
...
+            .iter()
+            .find(|field| &field.name == target_field)
+            .ok_or_else(|| {
+                let msg = format!(
+                    "field `{}` does not exist on type `{}`",
+                    target_field, target_type_name
+                );
+                invalid(object_type, &field.name, &msg)
+            })?;
+        // The field we are deriving from has to point back to us; as an
```
Comment (Jannis) [inline graph/src/data/graphql/validation.rs]: Nit: add an empty line above.

File: graph/src/data/graphql/validation.rs (hunk 1 above)
Comment (lutter) [reply graph/src/data/graphql/validation.rs]: I agree that that would be good to do, but I also do not want to do that as part of this PR which just builds on what's there already. I opened #1265 to track that.

**Comment by lutter** [comment]
Addressed review comments and rebased to latest master.

**Pull request merged**